File: hs/day2.hs
import qualified Data.ByteString as BS
import qualified Data.Text as DT
import Data.Text.Encoding(decodeUtf8)
import Data.Word

data Move = Rock | Paper | Scissors deriving(Show, Eq)

parseMove :: Char -> Move
parseMove 'A' = Rock
parseMove 'B' = Paper
parseMove 'C' = Scissors
parseMove 'X' = Rock
parseMove 'Y' = Paper
parseMove 'Z' = Scissors
parseMove _ = error "bad input"

valueOf :: Move -> Word64
valueOf Rock = 1
valueOf Paper = 2
valueOf Scissors = 3

findWinner :: Move -> Move
findWinner Rock = Paper
findWinner Paper = Scissors
findWinner Scissors = Rock

findLoser :: Move -> Move
findLoser Rock = Scissors
findLoser Paper = Rock
findLoser Scissors = Paper

part1 :: DT.Text -> Word64
part1 input = sum scores
    where
        lineScore line = score (parseMove $ DT.head line) (parseMove $ DT.last line)
        scores = lineScore <$> DT.lines input

score :: Move -> Move -> Word64
score them us = valueOf us +
    if us == them then 3
    else if us == findWinner them then 6
    else 0

part2 :: DT.Text -> Word64
part2 input = sum scores
    where
        theirs l = parseMove $ DT.head l
        lineScore l = case DT.last l of
            'X' -> valueOf . findLoser . theirs $ l
            'Y' -> (valueOf . theirs $ l) + 3
            'Z' -> (valueOf . findWinner . theirs $ l) + 6
            _ -> error "bad input"
        scores = lineScore <$> DT.lines input

main :: IO ()
main = do
    input <- decodeUtf8 <$> BS.readFile "../input/2022/day2.txt"
    print $ part1 input
    print $ part2 input

File: src/day2.rs
use aoc_runner_derive::aoc;

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
enum Move {
    Rock,
    Paper,
    Scissors,
}

impl From<u8> for Move {
    fn from(b: u8) -> Self {
        match b {
            0 => Move::Rock,
            1 => Move::Paper,
            2 => Move::Scissors,
            _ => panic!("bad input"),
        }
    }
}

fn value_of(m: Move) -> i64 {
    match m {
        Move::Rock => 1,
        Move::Paper => 2,
        Move::Scissors => 3,
    }
}

fn find_loser(m: Move) -> Move {
    match m {
        Move::Rock => Move::Scissors,
        Move::Paper => Move::Rock,
        Move::Scissors => Move::Paper,
    }
}

fn find_winner(m: Move) -> Move {
    match m {
        Move::Rock => Move::Paper,
        Move::Paper => Move::Scissors,
        Move::Scissors => Move::Rock,
    }
}

#[aoc(day2, part1)]
fn part1(input: &str) -> i64 {
    let mut score = 0i64;

    for line in input.lines() {
        let line = line.as_bytes();
        let (them, us): (Move, Move) = ((line[0] - b'A').into(), (line[2] - b'X').into());

        score += value_of(us);

        score += match (them, us) {
            (l, r) if l == r => 3,
            (l, r) if r == find_winner(l) => 6,
            (_, _) => 0,
        }
    }
    score
}

#[aoc(day2, part2)]
fn part2(input: &str) -> i64 {
    let mut score = 0i64;

    for line in input.lines() {
        let line = line.as_bytes();
        let them: Move = (line[0] - b'A').into();

        match line[2] {
            b'X' => score += value_of(find_loser(them)),
            b'Y' => score += value_of(them) + 3,
            b'Z' => score += value_of(find_winner(them)) + 6,
            _ => panic!("bad input"),
        }
    }
    score
}
